app/src/main/kotlin/at/bitfire/davdroid/sync/Syncer.kt

```diff
@@ -62,62 +64,112 @@ abstract class Syncer {
-
-    fun onPerformSync(
-        account: Account,
-        extras: Array<String>,
-        authority: String,
-        syncResult: SyncResult
-    ) {
-        logger.log(Level.INFO, "$authority sync of $account initiated", extras.joinToString(", "))
+    fun sync(provider: ContentProviderClient) {
+
+        // 0. resource specific preparations
+        beforeSync(provider)
+
+        // 1. find resource collections to be synced
+        val service = serviceRepository.getByAccountAndType(account.name, serviceType)
+        if (service != null)
+            for (remoteCollection in getSyncCollections(service.id))
+                remoteCollections[remoteCollection.url] = remoteCollection
+
+        // 2. update/delete local resources and determine new (unknown) remote collections
+        val newRemoteCollections = HashMap(remoteCollections)
+        for (localCollection in localCollections(provider)) {
+            val remoteCollection = remoteCollections[getUrl(localCollection)]
+            if (remoteCollection == null)
+                // Collection got deleted on server, delete obsolete local resource
+                delete(localCollection)
+            else {
+                // Collection exists locally, update local resource and don't add it again
+                update(localCollection, remoteCollection)
+                newRemoteCollections -= remoteCollection.url
+            }
+        }
+
+        // 3. create new local collections for newly found remote collections
+        for ((_, collection) in newRemoteCollections)
+            create(provider, collection)
+
+        // 4. sync local resources
+        for (localCollection in localSyncCollections(provider))
+            remoteCollections[getUrl(localCollection)]?.let { remoteCollection ->
+                syncCollection(provider, localCollection, remoteCollection)
+            }
+
+    }
+
+
+    /** All local collections of a specific type (calendar, address book, etc) */
+    abstract fun localCollections(provider: ContentProviderClient): List<CollectionType>
 
-        // use contacts provider for address books
-        val contentAuthority =
-            if (authority == context.getString(R.string.address_books_authority))
-                ContactsContract.AUTHORITY
-            else
-                authority
+    /** Sync enabled local collections of specific type */
+    abstract fun localSyncCollections(provider: ContentProviderClient): List<CollectionType>
 
-        val accountSettings by lazy { accountSettingsFactory.forAccount(account) }
-        val httpClient = lazy { HttpClient.Builder(context, accountSettings).build() }
+    abstract fun beforeSync(provider: ContentProviderClient)
 
-        // acquire ContentProviderClient of authority to be synced
+    abstract fun getSyncCollections(serviceId: Long): List<Collection>
+
+    abstract fun getUrl(localCollection: CollectionType): HttpUrl?
```

Comment:
Could we drop this and add an `abstract val LocalCollection.url get()=…` instead? Then we could get the `url` directly from `BaseSyncer` and wouldn't need `Syncer.getUrl` at all.